### PR TITLE
Prevent illegal states for mount configs

### LINKED ISSUE #45: Make invalid data for mount states impossible
- Username/Password either both or neither need to be present for `MongoDB`, `MarkLogic`
- Path should not be `Maybe` for `SparkHDFS` (use `rootDir` instead)

These are breaking changes, so will do them when we're next making a release.

### PR DESCRIPTION
**Pull request by garyb**
Resolves #45

### CODE REVIEW

File: src/Quasar/Mount/Common.purs [hunk 1]
```diff
@@ -40,14 +35,31 @@ extractHost (Just (URI.Authority _ hs)) =
...
+combineCredentials ∷ Credentials → String
+combineCredentials (Credentials { user, password })
+  | Str.null password = user
+  | otherwise = user <> ":" <> password
+
+extractCredentials ∷ Maybe URI.Authority → Maybe Credentials
+extractCredentials auth = do
+  userInfo ← (\(URI.Authority mui _) → mui) =<< auth
+  pure $ Credentials $
+    case Str.indexOf (Str.Pattern ":") userInfo of
```
Comment (kritzcreek) [inline src/Quasar/Mount/Common.purs:57]: https://pursuit.purescript.org/packages/purescript-strings/3.0.0/docs/Data.String#v:splitAt

File: src/Quasar/Mount/Common.purs [hunk 2]
```diff
@@ -40,14 +35,31 @@ extractHost (Just (URI.Authority _ hs)) =
...
+
+combineCredentials ∷ Credentials → String
+combineCredentials (Credentials { user, password })
+  | Str.null password = user
+  | otherwise = user <> ":" <> password
+
+extractCredentials ∷ Maybe URI.Authority → Maybe Credentials
+extractCredentials auth = do
+  userInfo ← (\(URI.Authority mui _) → mui) =<< auth
+  pure $ Credentials $
```
Comment (kritzcreek) [inline src/Quasar/Mount/Common.purs:56]: superfluous `$`

**Review by kritzcreek — APPROVED** [review]
Good from my side, added some suggestions

File: src/Quasar/Mount/Common.purs (hunk 1 above)
Comment (garyb) [reply src/Quasar/Mount/Common.purs:57]: I'm not sure `case flip Str.splitAt userInfo =<< Str.indexOf (Str.Pattern ":") userInfo of` is better than what's already there 😜

Comment (garyb) [reply src/Quasar/Mount/Common.purs:57]: Oh, yeah, there's `split` though.

Comment (garyb) [reply src/Quasar/Mount/Common.purs:57]: Nah, I think this is probably still better, as we do only want the first `:` to be split.

Comment (kritzcreek) [reply src/Quasar/Mount/Common.purs:57]: Hmm I meant use it to replace the take/drop usage in the `Just` branch, but you want to remove the `:` so you'd still end up with another drop.

**Pull request merged**